sklearn/metrics/_ranking.py

```diff
@@ -525,14 +526,21 @@ def _binary_clf_curve(y_true, y_score, pos_label=None, sample_weight=None):
         sample_weight = column_or_1d(sample_weight)
 
     # ensure binary classification if pos_label is not specified
+    # _determine_key_type(classes) == 'str' is required to avoid
+    # triggering a FutureWarning by calling np.array_equal(a, b)
+    # when elements in the two arrays are not comparable.
     classes = np.unique(y_true)
-    if (pos_label is None and
-        not (np.array_equal(classes, [0, 1]) or
-             np.array_equal(classes, [-1, 1]) or
-             np.array_equal(classes, [0]) or
-             np.array_equal(classes, [-1]) or
-             np.array_equal(classes, [1]))):
-        raise ValueError("Data is not binary and pos_label is not specified")
+    if (pos_label is None and (
+            _determine_key_type(classes) == 'str' or
```

Comment:
We can't use _determine_key_type here, because y_true can be floats (e.g., [0.0, 1.0]) and _determine_key_type will raise an error in this case.

Reply:
Since `classes = np.unique(y_true)`, is `classes.dtype.kind in {'O', 'U', 'S'}` good enough?

Reply:
I use ('U', 'S') in #15412, which is consistent with
https://github.com/scikit-learn/scikit-learn/blob/ff878d60aab59dadf3bc1ee2e914fd038662a71a/sklearn/preprocessing/_encoders.py#L132

Reply:
I'll add a test for this case (float labels) and update the code accordingly.